lib/absinthe/phase/operation/input.ex

```diff
@@ -0,0 +1,129 @@
+  - The `item` field's `category` argument struct would be modified to have a
+    `provided_value` of `"Things"`.
+
+  If, the following was used (note no `"id"` variable value is given):
+
+  ```
+  run(blueprint, %{operation_name: "Item", variables: %{}})
+  ``
+
+  - The variable definition for `id` would have an error record in its `errors`
+    field.
+  - The `item` field's `category` argument struct would be modified to have a
+    `provided_value` of `"Things"`.
+  """
+
+  alias Absinthe.{Blueprint, Phase}
+
+  @error_not_provided "value not provided"
+
+  @spec run(Blueprint.t, %{operation_name: nil | String.t, variables: %{String.t => any}}) :: {:ok, Blueprint.t}
+  def run(input, options) do
+    acc = Map.put(options, :active_operation, false)
+    {node, _} = Blueprint.Mapper.prewalk(input, acc, &handle_node/2)
+    {:ok, node}
+  end
+
+  @spec handle_node(Blueprint.node_t, map) :: {Blueprint.node_t, map}
+  # Matching operation: Mark active, process variable definitions
+  defp handle_node(%Blueprint.Operation{name: name} = node, %{operation_name: name} = acc) do
+    # Let children know they're in the right operation
+    acc = %{acc | active_operation: true}
+    # We need to traverse the variable definitions first, so that we can harvest
+    # any default values and put them in the `variables` field of our accumulator
+    {variable_definitions, acc} = Enum.map_reduce(node.variable_definitions, acc, &process_variable_definition/2)
+    {
+      struct(node, variable_definitions: variable_definitions),
+      acc
+    }
+  end
+  # Non-matching operation: Mark inactive
+  defp handle_node(%Blueprint.Operation{} = node, acc) do
+    {node, %{acc | active_operation: false}}
+  end
+  # Argument using a variable: Set provided value
+  defp handle_node(%Blueprint.Input.Argument{value: %Blueprint.Input.Variable{name: variable_name}} = node, %{active_operation: true} = acc) do
+    {
+      struct(node, provided_value: Map.get(acc.variables, variable_name)),
+      acc
+    }
+  end
+  # Argument not using a variable: Set provided value from the literal value
+  defp handle_node(%Blueprint.Input.Argument{} = node, %{active_operation: true} = acc) do
+    {
+      struct(node, provided_value: Blueprint.Input.unwrap(node.value)),
+      acc
+    }
+  end
+
+  # All other nodes: return unchanged
+  defp handle_node(node, acc) do
+    {node, acc}
+  end
+
+  defp process_variable_definition(node, acc) do
+    case {node.type, Map.get(acc.variables, node.name)} do
+      # If it's non-null and there's no value, add an error
+      {%Blueprint.NonNullType{}, nil} ->
+        {
+          update_in(node.errors, &[error(@error_not_provided) | &1]),
+          acc
+        }
+      # If it's optional, assign its (bare, unwrapped) default value to the
+      # variables map
+      {_, nil} ->
+        {
+          node,
+          %{acc | variables: Map.put(acc.variables, node.name, Blueprint.Input.unwrap(node.default_value))}
```

Comment:
This won't be bare in the same sense that a value in JSON is bare. Default values would still look like `%Input.Integer{value: 4}` whereas variables from JSON are just `4`

Also, unwrapping eliminates any list wrapping which we definitely don't want to do here.